scope_benchmark_test.go

```diff
@@ -139,6 +141,75 @@ func BenchmarkScopeTaggedNoCachedSubscopes(b *testing.B) {
 	}
 }
 
+func BenchmarkScopeTaggedRandomCachedSubscopesParallel(b *testing.B) {
+	root, _ := NewRootScope(ScopeOptions{
+		Prefix:   "funkytown",
+		Reporter: NullStatsReporter,
+		Tags: map[string]string{
+			"style":     "funky",
+			"hair":      "wavy",
+			"jefferson": "starship",
+		},
+	}, 0)
+
+	b.ResetTimer()
+
+	index := int64(0)
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			n := atomic.AddInt64(&index, 1)
+			value := strconv.Itoa(int(n))
+
+			root.Tagged(map[string]string{
+				"foo": value,
+				"baz": value,
+				"qux": value,
+			})
+		}
+	})
+}
+
+func BenchmarkScopeTaggedRandomCachedSubscopesParallelRandom(b *testing.B) {
+	percentageCached := 5
+	root, _ := NewRootScope(ScopeOptions{
+		Prefix:   "funkytown",
+		Reporter: NullStatsReporter,
+		Tags: map[string]string{
+			"style":     "funky",
+			"hair":      "wavy",
+			"jefferson": "starship",
+		},
+	}, 0)
+
+	cachedMap := map[string]string{
+		"foo": "any",
+		"baz": "any",
+		"qux": "any",
+	}
+	root.Tagged(cachedMap)
+
+	b.ResetTimer()
+
+	index := int64(0)
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			if rand.Intn(100) < percentageCached {
+				root.Tagged(cachedMap)
+				continue
+			}
+
+			n := atomic.AddInt64(&index, 1)
+			value := strconv.Itoa(int(n))
+
+			root.Tagged(map[string]string{
```

Comment:
Shouldn't we be storing the result of this somewhere, so as to not have the code optimized away? 

> Before concluding I wanted to highlight that to be completely accurate, any benchmark should be careful to avoid compiler optimisations eliminating the function under test and artificially lowering the run time of the benchmark.

https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go

Reply:
I could do that, but I actually checked this was being called through a cpu profile.

<img width="442" alt="Screen Shot 2022-10-10 at 12 26 45 PM" src="https://user-images.githubusercontent.com/3793727/194938931-e81cfda1-7175-4ac7-8004-636be30ef205.png">

Do you still want me to store it? I would need to also store in multiple buckets so that it doesn't become a bottleneck for the benchmark.

Reply:
Fair point r.e. the parallelization--so long as it's not getting optimized away I think it's good. Add a comment to the code documenting what you just mentioned so that others don't come in with the same idea/comment.